allennlp/data/semparse/type_declarations/type_declaration.py

```
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
+    However, if the complex type contains ANY_TYPE, we substitute it with all possible basic types. If it
+    contains placeholders, the substituyions need to be constrained. For example, for <#1,#1>, <e,r> is not a
+    valid substitution.
+    """
+    valid_actions: Dict[str, Set[str]] = defaultdict(set)
+
+    complex_types = set()
+    for name, alias in name_mapping.items():
+        if not alias in type_signatures:
+            continue
+        name_type = type_signatures[alias]
+        # Type to terminal productions.
+        for substituted_type in _substitute_any_type(name_type, basic_types):
+            valid_actions[str(substituted_type)].add(_make_production_string(substituted_type, name))
+        # Keeping track of complex types.
+        if isinstance(name_type, ComplexType) and name_type != ANY_TYPE:
+            complex_types.add(name_type)
+
+    for complex_type in complex_types:
+        if isinstance(complex_type, PlaceholderType):
+            if complex_type.first == ANY_TYPE:
+                if isinstance(complex_type.first, BasicType):
+                    for basic_type in basic_types:
+                        application_type = complex_type.get_application_type(basic_type)
+                        valid_actions[str(application_type)].add(_make_production_string(application_type,
+                                                                                         [complex_type,
+                                                                                          basic_type]))
+                        for head, production in _get_complex_type_productions(application_type):
+                            valid_actions[head].add(production)
+                else:
+                    # This means complex_type.first is ComplexType(ANY_TYPE, ANY_TYPE)
+                    # TODO(pradeep): Assuming this is a reverse type. That is the only type where the
```

Instead of a `TODO`, can you check that `complex_type` actually _is_ a reverse type, and raise a `NotImplementedError` otherwise?  Better to crash than to fail silently in cases where you're explicitly not supporting something.

---

Done.